File: src/jastudio_src/jastudio/note/vocabulary/related_vocab/related_vocab_data.py
from __future__ import annotations

from typing import TYPE_CHECKING, override

from autoslot import Slots  # pyright: ignore[reportMissingTypeStubs]
from jaslib.sysutils.lazy import Lazy
from jastudio.note.vocabulary.related_vocab.related_vocab_data_serializer import RelatedVocabDataSerializer

if TYPE_CHECKING:
    from jastudio.note.notefields.auto_save_wrappers.value_wrapper import ValueWrapper
    from typed_linq_collections.collections.q_set import QSet

class RelatedVocabData(Slots):
    serializer: Lazy[RelatedVocabDataSerializer] = Lazy(RelatedVocabDataSerializer)

    def __init__(self, ergative_twin: str, derived_from: ValueWrapper[str], perfect_synonyms: QSet[str], similar: QSet[str], antonyms: QSet[str], confused_with: QSet[str], see_also: QSet[str]) -> None:
        self.ergative_twin: str = ergative_twin
        self.derived_from: ValueWrapper[str] = derived_from

        self.synonyms: QSet[str] = similar
        self.perfect_synonyms: QSet[str] = perfect_synonyms
        self.antonyms: QSet[str] = antonyms
        self.confused_with: QSet[str] = confused_with
        self.see_also: QSet[str] = see_also

    @override
    def __repr__(self) -> str: return f"RelatedVocabData(ergative_twin={self.ergative_twin}, derived_from={self.derived_from}, perfect_synonyms={self.perfect_synonyms}, synonyms={self.synonyms}, antonyms={self.antonyms}, confused_with={self.confused_with}, see_also={self.see_also})"

File: src/src_dotnet/JAStudio.Core/Note/Vocabulary/RelatedVocab/RelatedVocabData.cs
using System;
using System.Collections.Generic;
using System.Linq;
using JAStudio.Core.Note.NoteFields.AutoSaveWrappers;

namespace JAStudio.Core.Note.Vocabulary.RelatedVocab;

public class RelatedVocabData
{
    private static readonly Lazy<RelatedVocabDataSerializer> SerializerInstance = new(() => new RelatedVocabDataSerializer());

    public static RelatedVocabDataSerializer Serializer() => SerializerInstance.Value;

    public string ErgativeTwin { get; set; }
    public ValueWrapper<string> DerivedFrom { get; set; }
    public HashSet<string> Synonyms { get; set; }
    public HashSet<string> PerfectSynonyms { get; set; }
    public HashSet<string> Antonyms { get; set; }
    public HashSet<string> ConfusedWith { get; set; }
    public HashSet<string> SeeAlso { get; set; }

    public RelatedVocabData(
        string ergativeTwin,
        ValueWrapper<string> derivedFrom,
        HashSet<string> perfectSynonyms,
        HashSet<string> similar,
        HashSet<string> antonyms,
        HashSet<string> confusedWith,
        HashSet<string> seeAlso)
    {
        ErgativeTwin = ergativeTwin;
        DerivedFrom = derivedFrom;
        Synonyms = similar;
        PerfectSynonyms = perfectSynonyms;
        Antonyms = antonyms;
        ConfusedWith = confusedWith;
        SeeAlso = seeAlso;
    }

    public override string ToString()
    {
        return $"RelatedVocabData(ergative_twin={ErgativeTwin}, derived_from={DerivedFrom}, " +
               $"perfect_synonyms={string.Join(",", PerfectSynonyms)}, synonyms={string.Join(",", Synonyms)}, " +
               $"antonyms={string.Join(",", Antonyms)}, confused_with={string.Join(",", ConfusedWith)}, " +
               $"see_also={string.Join(",", SeeAlso)})";
    }
}
